sharedbuffers/mapped_struct.py

```diff
@@ -2667,18 +2904,29 @@ def init(self, slot_types = None, slot_keys = None, alignment = 8, pack_buffer_s
 
     @cython.ccall
     @cython.locals(has_bitmap = cython.ulonglong, none_bitmap = cython.ulonglong, present_bitmap = cython.ulonglong,
-        i = int)
+        i = int, get_values = cython.bint)
     @cython.returns(tuple)
-    def _get_bitmaps(self, obj):
+    def _get_bitmaps(self, obj, get_values=False):
         has_bitmap = 0
         none_bitmap = 0
+        if get_values:
+            values = []
         for i,slot in enumerate(self.slot_keys):
-            if hasattr(obj, slot):
+            try:
+                val = getattr(obj, slot)
+            except AttributeError:
```

Comment:
is this actually better than doing a ```hasattr``` test ? I'd have thought that exception handling was always slower.

Reply:
Exception handling in Cython is quite fast. That said, I haven't timed it on basic attributes. But it's most likely faster due to another subtle reason: hasattr on properties (ie: all proxied properties) has to invoke the getter and check if it rases AttributeError. That is, this exact code, only it discards the value if it succeeds. This is faster on properties because it avoids one whole invocation to the getter, which can be quite expensive.